lib/src/connection.dart

```diff
@@ -495,3 +503,122 @@ class Notification {
 }
+
+class PostgreSQLConnectionPool {
+  PostgreSQLConnectionPool(this.size, this.host, this.port, this.databaseName,
+      {this.username: null,
+        this.password: null,
+        this.timeoutInSeconds: 30,
+        this.timeZone: "UTC",
+        this.useSSL: false,
+        int maxRetryIntervalInSeconds = 5})
+  : _maxRetryInterval = new Duration(seconds: maxRetryIntervalInSeconds);
+
+  final int size;
+  final String host;
+  final int port;
+  final String databaseName;
+  final String username;
+  final String password;
+  final int timeoutInSeconds;
+  final String timeZone;
+  final bool useSSL;
+  final Duration _maxRetryInterval;
+  final Completer _done = new Completer();
+
+  bool get _isClosed => _done.isCompleted;
+
+  var _connections = new Set<PostgreSQLConnection>();
+  int _countFailedConnected = 0;
+
+  /// Get opened connection from pool
+  PostgreSQLConnection get connection {
+    if(_isClosed) {
+      throw new PostgreSQLException(
+          "Attempting to get connection, but pool is closed.");
+    }
+    if(_connections.length == 0) {
+      throw new PostgreSQLException(
+          "Attempting to get connection, but has not available connections.");
+    }
+    PostgreSQLConnection selectedConnection = _connections.first;
+    int maxLengthQueryQueue() => selectedConnection._queryQueue.length;
+    for(var connection in _connections.skip(1)) {
+      if(connection._queryQueue.length < maxLengthQueryQueue()) {
+        selectedConnection = connection;
+      }
+    }
+    return selectedConnection;
+  }
+
+  /// Open connections with a PostgreSQL database.
+  Future open() async {
+    if(_isClosed) {
+      throw new PostgreSQLException(
+          "Attempting to reopen a closed connectionPool. Create a new instance instead.");
+    }
+    var newConnections = new List.generate(size, (_) => _createConnection());
+    await Future.wait(newConnections.map(_aliveConnection));
+  }
+
+  PostgreSQLConnection _createConnection() {
+    var connection = new PostgreSQLConnection(host, port, databaseName,
+        username: username,
+        password: password,
+        timeoutInSeconds: timeoutInSeconds,
+        timeZone: timeZone,
+        useSSL: useSSL);
+    connection.done.then((_) async {
+      if(_isClosed) {
+        return;
+      }
+
+      _connections.remove(connection);
+      var delayed = new Future.delayed(_getRetryDelayedDuration());
+      await Future.any([_done.future, delayed]);
+
+      if(_isClosed) {
+        return;
+      }
+
+      await _aliveConnection(_createConnection());
+    });
+    return connection;
+  }
+
+  Duration _getRetryDelayedDuration() {
+    int exponent = _countFailedConnected ~/ size;
+    var result = new Duration(milliseconds: pow(2, exponent));
+
+    if(result > _maxRetryInterval) {
+      result = _maxRetryInterval;
+    }
+    return result;
+  }
+
+  Future _aliveConnection(PostgreSQLConnection connection) async {
+    try {
+      await connection.open();
```

Comment:
"Recreation" test will ensure that you can make 5 queries after shutting down five connections, but it doesn't ensure that the pool size returns to normal; technically that test could pass if only 1 connection was recreated.